### PR TITLE
[Example] Pytorch Seal example

### PR DESCRIPTION
**Pull request by Smilexuhc**
## Description
Pytorch Implementation of Seal in paper *Link Prediction Based on Graph Neural Networks*.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/seal/gcn_layers.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
+# Most codes from dgl.nn.pytorch
+# Adjust codes in forward for utilizing edge weight in message passing.
```
Comment (yzh119) [inline examples/pytorch/seal/gcn_layers.py:2]: The `SAGEConv` and `GraphConv` module has already supported `edge_weight`.

File: examples/pytorch/seal/sampler.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+    def __len__(self):
+        """Return the number of batches of the data loader."""
+        return len(self.dataloader)
+
+    def __iter__(self):
+        return iter(self.dataloader)
+
+
+class PosNegEdgesGenerator(object):
```
Comment (yzh119) [inline examples/pytorch/seal/sampler.py:58]: Should this Generator be accelerated with `DataLoader`?

File: examples/pytorch/seal/utils.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        src(int): node id of one of src node in new subgraph
+        dst(int): node id of one of dst node in new subgraph
+    Returns:
+        z(Tensor): node labeling tensor
+    """
+    adj = subgraph.adj().to_dense().numpy()
+    src, dst = (dst, src) if src > dst else (src, dst)
+
+    idx = list(range(src)) + list(range(src + 1, adj.shape[0]))
+    adj_wo_src = adj[idx, :][:, idx]
```
Comment (yzh119) [inline examples/pytorch/seal/utils.py:115]: Could you please elaborate more on `adj_wo_src` and `adj_wo_dst`?

File: examples/pytorch/seal/sampler.py (hunk 2 above)
Comment (Smilexuhc) [reply examples/pytorch/seal/sampler.py:58]: I think there is no need for accelerating the Generator, since generating all the 2500000 negative edges cost less than 0.1s.

File: examples/pytorch/seal/utils.py (hunk 3 above)
Comment (Smilexuhc) [reply examples/pytorch/seal/utils.py:115]: ok

File: examples/pytorch/seal/gcn_layers.py (hunk 1 above)
Comment (Smilexuhc) [reply examples/pytorch/seal/gcn_layers.py:2]: Ok

File: examples/pytorch/seal/utils.py (hunk 3 above)
Comment (Smilexuhc) [reply examples/pytorch/seal/utils.py:115]: The codes are mainly from the author. I will simplify the codes for more readable and efficient.

File: examples/pytorch/seal/logger.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            logging.basicConfig(level=log_level,
+                                format="%(asctime)s %(levelname)s: %(message)s",
+                                datefmt='%Y-%m-%d-%H:%M',
+                                handlers=[
+                                    logging.FileHandler(log_name, mode='w'),
+                                    logging.StreamHandler()
+                                ])
+            logging.info('Start Logging')
+            logging.info('Log file path: {}'.format(log_name))
+
```
Comment (sneakerkg) [inline examples/pytorch/seal/logger.py:46]: Remove this empty line.

File: examples/pytorch/seal/main.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
+import dgl
```
Comment (sneakerkg) [inline examples/pytorch/seal/main.py:1]: Usually we will follow some orders on import:
1.Original python packages, like os, sys;
2.Commonly used packages, like numpy, scipy, tqdm, etc;
3.The specific packages you are using, like dgl;
4.Code in other files in this example.

File: examples/pytorch/seal/README.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ - NumTestEdges: 146329
+ 
+Dependencies
+--------------------------------
+
+- python 3.6+
+- Pytorch 1.5.0+
+- pandas
+- tqdm
+- scipy
```
Comment (sneakerkg) [inline examples/pytorch/seal/README.md:30]: Missing ogb.

File: examples/pytorch/seal/main.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+from utils import load_ogb_dataset, evaluate_hits
+import torch
+from sampler import SEALData, SEALDataLoader
+from model import GCN, DGCNN
+import numpy as np
+from logger import LightLogging
+import time
+
+
+def train(model, dataloader, loss_fn, optimizer, device, num_graphs=32, total_graphs=None):
```
Comment (sneakerkg) [inline examples/pytorch/seal/main.py:15]: Train and eval logic are similar with the official implementation. Please add a comment on that, similar as: https://github.com/dmlc/dgl/blob/master/examples/pytorch/pointcloud/pointnet/pointnet2.py#L10

File: examples/pytorch/seal/main.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    node_attribute = seal_data.ndata['feat']
+    edge_weight = seal_data.edata['edge_weight'].float()
+
+    train_data = seal_data('train')
+    val_data = seal_data('valid')
+    test_data = seal_data('test')
+
+    train_graphs = len(train_data.graph_list)
+
+    # Set data loader
```
Comment (sneakerkg) [inline examples/pytorch/seal/main.py:91]: The model will be out with 0.6 release. In that case we may only need the GraphDataLoader solution.

File: examples/pytorch/seal/main.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+            test_metric = evaluate_hits(args.dataset, test_pos_pred, test_neg_pred, args.hits_k)
+            evaluate_time = time.time()
+            print_fn("Epoch-{}, train loss: {:.4f}, hits@{}: val-{:.4f}, test-{:.4f}, "
+                     "cost time: train-{:.1f}s, total-{:.1f}s".format(epoch, loss, args.hits_k, val_metric, test_metric,
+                                                                      train_time - start_time,
+                                                                      evaluate_time - start_time))
+            summary_val.append(val_metric)
+            summary_test.append(test_metric)
+
+    # summary_val = np.array(summary_val)
```
Comment (sneakerkg) [inline examples/pytorch/seal/main.py:165]: Remove comments.

File: examples/pytorch/seal/model.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl.nn.pytorch import SortPooling, SumPooling
+from gcn_layers import GraphConv, SAGEConv
```
Comment (sneakerkg) [inline examples/pytorch/seal/model.py:5]: As suggested by Zihao, please use dgl.nn modules here.

File: examples/pytorch/seal/README.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+ How to run example files
+--------------------------------
+In the seal_dgl folder    
+run on cpu:  
+```shell script
+python main.py --gpu_id=-1 --subsample_ratio=0.1
```
Comment (sneakerkg) [inline examples/pytorch/seal/README.md:37]: The model has a lot of options, like gcn or dgcnn. Seems not reflected in the CMD. Are we all using default setting? If so, maybe we should make sure other options also works, at least not crash.

File: examples/pytorch/seal/model.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+    def forward(self, g, z, node_id=None, edge_id=None):
+        """
+        Args:
+            g(DGLGraph): the graph
+            z(Tensor): node labeling tensor, shape [N, 1]
+            node_id(Tensor, optional): node id tensor, shape [N, 1]
+            edge_id(Tensor, optional): edge id tensor, shape [E, 1]
+        Returns:
+            x(Tensor): output tensor
+
```
Comment (sneakerkg) [inline examples/pytorch/seal/model.py:92]: Remove empty line.

File: examples/pytorch/seal/sampler.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+    Data Loader of SEAL
+    Attributes:
+        dataset(Dataset): dataset
+        batch_size(int): size of batch
+    """
+
+    def __init__(self, dataset, batch_size, num_workers=1, shuffle=True,
+                 drop_last=False, pin_memory=False):
+        self.total_graphs = len(dataset)
+        self.dataloader = DataLoader(dataset=dataset, collate_fn=self._collate, batch_size=batch_size, shuffle=shuffle,
```
Comment (sneakerkg) [inline examples/pytorch/seal/sampler.py:40]: Why not inherit from Dataloader?

File: examples/pytorch/seal/sampler.py [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+    def __call__(self, split_type):
+
+        if split_type == 'train':
+            subsample_ratio = self.subsample_ratio
+        else:
+            subsample_ratio = 1
+
+        pos_edges = self.split_edge[split_type]['edge']
+        if split_type == 'train':
+            g = add_self_loop(self.g)
```
Comment (sneakerkg) [inline examples/pytorch/seal/sampler.py:86]: Why only train need self_loop?

File: examples/pytorch/seal/sampler.py [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+            eids = g.edge_ids(pos_edges[:, 0], pos_edges[:, 1])
+            neg_edges = torch.stack(self.neg_sampler(g, eids), dim=1)
+        else:
+            neg_edges = self.split_edge[split_type]['edge_neg']
+        pos_edges = self.subsample(pos_edges, subsample_ratio).long()
+        neg_edges = self.subsample(neg_edges, subsample_ratio).long()
+
+        if self.return_type == 'split':
```
Comment (sneakerkg) [inline examples/pytorch/seal/sampler.py:94]: When to use split and when to use combine?

File: examples/pytorch/seal/sampler.py [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+        elif self.return_type == 'combine':
+            edges = torch.cat([pos_edges, neg_edges])
+            labels = torch.cat([torch.ones(pos_edges.size(0), 1), torch.zeros(neg_edges.size(0), 1)])
+            if self.shuffle:
+                perm = torch.randperm(edges.size(0))
+                edges = edges[perm]
+                labels = labels[perm]
+            return edges, labels
+
+    def subsample(self, edges, subsample_ratio):
```
Comment (sneakerkg) [inline examples/pytorch/seal/sampler.py:105]: Add comments.

File: examples/pytorch/seal/sampler.py [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+        self.save_dir = save_dir
+        self.print_fn = print_fn
+
+        self.generator = PosNegEdgesGenerator(g=self.g,
+                                              split_edge=split_edge,
+                                              neg_samples=neg_samples,
+                                              subsample_ratio=subsample_ratio,
+                                              shuffle=shuffle,
+                                              return_type='combine')
+        if use_coalesce:
```
Comment (sneakerkg) [inline examples/pytorch/seal/sampler.py:254]: When to use? Add comments.

**Review by sneakerkg — CHANGES_REQUESTED** [review]

File: examples/pytorch/seal/sampler.py (hunk 14 above)
Comment (sneakerkg) [reply examples/pytorch/seal/sampler.py:86]: Add comments.

**Comment by sneakerkg** [comment]
Also add paper information in: examples/README.md

**Comment by yzh119** [comment]
btw, `to_simple` has supported coalescing graph in #2600 , you can have a try.

**Comment by Smilexuhc** [comment]
> btw, `to_simple` has supported coalescing graph in #2600 , you can have a try.

Get.

File: examples/pytorch/seal/utils.py [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+    else:
+        graph = dgl.to_simple(graph)
+
+    src, dst = graph.edges()
+    graph_df = pd.DataFrame({'src': src, 'dst': dst})
+    graph_df = pd.merge(graph_df, tmp, how='left', on=['src', 'dst'])
+    graph.edata['edge_weight'] = torch.from_numpy(graph_df['edge_weight'].values).unsqueeze(1)
+
+    graph.edata.pop('count')
+    return graph
```
Comment (yzh119) [inline examples/pytorch/seal/utils.py:93]: I suppose we can remove this function now.

File: examples/pytorch/seal/model.py [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+        """
+
+        z_emb = self.z_embedding(z)
+
+        # if z_emb.ndim == 3:  # in case z has multiple integer labels
```
Comment (yzh119) [inline examples/pytorch/seal/model.py:95]: Remove comments.

File: examples/pytorch/seal/sampler.py [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+        #     if subgraph.has_edges_between(u_id, v_id):
+        #         link_id = subgraph.edge_ids(u_id, v_id, return_uv=True)[2]
+        #         subgraph.remove_edges(link_id)
+        #         eids = eids[subgraph.edata[EID]]
+        #     if subgraph.has_edges_between(v_id, u_id):
+        #         link_id = subgraph.edge_ids(v_id, u_id, return_uv=True)[2]
+        #         subgraph.remove_edges(link_id)
+        #         eids = eids[subgraph.edata[EID]]
+        #     subgraph.ndata[NID] = nids
+        #     subgraph.edata[EID] = eids
```
Comment (yzh119) [inline examples/pytorch/seal/sampler.py:164]: Remove comments.

File: examples/pytorch/seal/README.md [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+ - NumNodes: 235868
+ - NumEdges: 2358104
+ - NumNodeFeats: 128
+ - NumEdgeWeights: 1
+ - NumValidEdges: 160084
+ - NumTestEdges: 146329
+ 
+Dependencies
+--------------------------------
+
```
Comment (yzh119) [inline examples/pytorch/seal/README.md:25]: Add `dgl 0.6+`

**Review by yzh119 — COMMENTED** [review]
I'll merge this PR after you address @sneakerkg 's concerns.

File: examples/pytorch/seal/utils.py (hunk 18 above)
Comment (Smilexuhc) [reply examples/pytorch/seal/utils.py]: removed

File: examples/pytorch/seal/sampler.py (hunk 13 above)
Comment (Smilexuhc) [reply examples/pytorch/seal/sampler.py]: Has replaced it with dgl.dataloading.GraphDataLoader()

**Review by yzh119 — APPROVED** [review]
LGTM

**Review by sneakerkg — APPROVED** [review]
